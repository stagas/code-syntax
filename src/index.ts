export * from './element'
export * from './syntax'
export * from './util'
