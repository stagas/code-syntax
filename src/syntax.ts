import { escape } from 'html-escaper'

export type RegExpMap = Record<string, RegExpMapped>
export type RegExpMapped = RegExp & { map: RegExpMap; keys: Set<string> }

export interface SyntaxDefinition {
  [k: string]: RegExp | [RegExp, SyntaxDefinition]
}

export type SyntaxOrImport = SyntaxDefinition | Promise<{ default: SyntaxDefinition }>

/**
 * Compiles a syntax definition.
 *
 * ```js
 * const r = await compile({
 *   foo: /[a-z]/,
 *   bar: /[0-9]/,
 * })
 * ```
 *
 * @param def The syntax definition to compile. Can be a promise returned by `import()`.
 */
export const compile = async (
  def: SyntaxOrImport,
  keys: Set<string> = new Set(),
): Promise<RegExpMapped> => {
  // try to extract the definitions if we're given either module or promise
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const R = (await ((await (def as any)?.default) ?? def))?.default ?? def
  if (!R) throw new ReferenceError('Could not find definition for language')
  const map: RegExpMap = {}
  const parts = []
  // eslint-disable-next-line prefer-const
  for (let [name, value] of Object.entries(R)) {
    keys.add(name)
    if (Array.isArray(value)) {
      map[name] = await compile(value[1], keys)
      value = value[0]
    }
    parts.push(`(?<${name}>${(value as unknown as RegExp)?.source ?? value})`)
  }
  const regexp = RegExp(parts.join('|'), 'gm') as RegExpMapped
  regexp.map = map
  regexp.keys = keys
  return regexp
}

/**
 * Syntax highlights a string as html with the given syntax.
 *
 * ```js
 * const regexp = await compile({
 *   foo: /[a-z]+/,
 *   bar: [
 *     /[0-9]+/,
 *     {
 *       bar: /[0-5]+/,
 *     },
 *   ],
 * })
 * const html = syntax(regexp, 'hello 123 789 world')
 * ```
 *
 * @param regexp The syntax definition returned by `compile()`
 * @param s The string to highlight.
 */
export const syntax = (regexp: RegExpMapped | RegExp, s: string): string =>
  s.replace(RegExp('(' + regexp.source + ')|(?<catchbad>[&<>"\'])', 'gm'), (...args: unknown[]) => {
    const groups = args.pop() as RegExpMatchArray
    const entries = Object.entries(groups).filter(e => e[1] != null)
    if (!entries.length) return ''
    const [tag, value] = entries[0]
    if (tag === 'catchbad') return escape(value)
    if ((regexp as RegExpMapped).map) {
      const sub = (regexp as RegExpMapped).map[tag]
      if (sub) return `<span class="${tag}-outer">${syntax(sub, value)}</span>`
    }
    return `<span class="${tag}">${escape(value)}</span>`
  })
