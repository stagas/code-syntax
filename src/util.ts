import { CodeSyntaxElement } from '.'

export interface PatchOptions {
  theme: string
  getLanguage: (el: Element) => string
}

export const patchElements = (
  elements: Element[],
  {
    theme = 'monokai',
    getLanguage = el =>
      [...el.classList]
        .filter(x => x.startsWith('language-'))[0]
        .split('-')
        .pop()!,
  } = {} as Partial<PatchOptions>
) =>
  elements.forEach(oldEl => {
    const newEl = new CodeSyntaxElement()
    newEl.textContent = oldEl.textContent
    newEl.setAttribute('theme', theme)
    newEl.setAttribute('language', getLanguage(oldEl))
    oldEl.replaceWith(newEl)
  })

export const patchPreCodeElements = (opts: Partial<PatchOptions>) =>
  patchElements(
    [...document.querySelectorAll('code')].filter(el => el.parentNode?.nodeName === 'PRE'),
    opts
  )
