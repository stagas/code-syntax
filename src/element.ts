import $ from 'sigl'

import { escape } from 'html-escaper'
import { compile, RegExpMapped, syntax as syntaxHighlight, SyntaxDefinition, SyntaxOrImport } from './syntax'

export type { RegExpMapped, SyntaxDefinition, SyntaxOrImport }

export const languages: Record<string, SyntaxOrImport> = {}

const compiled = new Map<object, { style: CSSStyleSheet; regexp: RegExpMapped }>()

const style = /*css*/ `
:host {
  contain: layout style paint;
  display: block;
}
* {
  contain: layout style paint;
  text-rendering: optimizeSpeed;
  color: var(--color);
}
*::selection {
  background: var(--selection) !important;
}
pre,
code {
  margin: 0;
  text-rendering: optimizeSpeed;
  font-family: inherit;
}
slot {
  display: none;
}`

const keyStyle = (x: string) => /*css*/ `
.${x} {
  color: var(--${x});
  font-style: var(--${x}-style);
  font-weight: var(--${x}-weight);
}`

export interface CodeSyntaxElement extends $.Element<CodeSyntaxElement> {}

@$.element()
export class CodeSyntaxElement extends HTMLElement {
  root = $.shadow(this, /*html*/ `<style>${style}</style><slot></slot><pre><code></code></pre>`)

  @$.attr() language = $.String
  @$.attr() theme = $.String

  syntax?: SyntaxOrImport
  codeRef?: HTMLElement
  codeHTML?: string
  codeRawText?: string
  highlight?: (s: string) => string
  onTextContent?: (textContent: string) => void

  mounted($: this['$']) {
    $.codeRef = $.query('code')

    // prevents unstyled content from showing until attributes
    // have arrived and the actual syntax has compiled
    setTimeout(() => {
      if ($.highlight == null) $.highlight = escape
    }, 100)

    $.effect(({ language }) => {
      if (language in languages) $.syntax = languages[language]
    })

    $.effect(({ root }) =>
      $.onTextChange(root, text => {
        $.codeRawText = text
      })
    )

    $.effect(async ({ root, syntax }) => {
      let { style, regexp } = compiled.get(syntax) ?? {}
      if (!regexp || !style) {
        regexp = await compile(syntax)
        style = new CSSStyleSheet()
        await style.replace([...regexp.keys].map(keyStyle).join(''))
        compiled.set(syntax, { style, regexp })
      }
      root.adoptedStyleSheets = [style]
      $.highlight = (s: string) => syntaxHighlight(regexp!, s)
    })

    $.codeHTML = $.reduce(({ codeRawText, highlight }) => highlight(codeRawText))

    $.effect(({ codeRef, codeHTML }) => {
      codeRef.innerHTML = codeHTML
    })
  }
}
