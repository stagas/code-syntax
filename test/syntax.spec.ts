import { compile, syntax } from '../src/syntax'

describe('compile', () => {
  it('simple object single rule', async () => {
    const r = await compile({
      foo: /[a-z]/,
    })
    expect(r.source).toEqual('(?<foo>[a-z])')
    expect([...r.keys!]).toEqual(['foo'])
  })

  it('promise module', async () => {
    const r = await compile(
      Promise.resolve({
        default: {
          foo: /[a-z]/,
        },
      })
    )
    expect(r.source).toEqual('(?<foo>[a-z])')
    expect([...r.keys!]).toEqual(['foo'])
  })

  it('simple object multiple rules', async () => {
    const r = await compile({
      foo: /[a-z]/,
      bar: /[0-9]/,
    })
    expect(r.source).toEqual('(?<foo>[a-z])|(?<bar>[0-9])')
    expect([...r.keys!]).toEqual(['foo', 'bar'])
  })

  it('simple object nested rules', async () => {
    const r = await compile({
      foo: /[a-z]/,
      bar: [
        /[0-9]/,
        {
          bar: /[0-5]/,
        },
      ],
    })
    expect(r.source).toEqual('(?<foo>[a-z])|(?<bar>[0-9])')
    expect([...r.keys!]).toEqual(['foo', 'bar'])
    expect(r.map!['bar'].source).toEqual('(?<bar>[0-5])')
  })
})

describe('syntax', () => {
  it('simple named group regexp', () => {
    const regexp = /(?<foo>[a-z]+)/g
    const result = syntax(regexp, 'hello 123 world')
    expect(result).toMatchSnapshot()
  })

  it('multiple named group regexp', () => {
    const regexp = /(?<foo>[a-z]+)|(?<bar>[0-9]+)/g
    const result = syntax(regexp, 'hello 123 world')
    expect(result).toMatchSnapshot()
  })

  it('mapped regexp', async () => {
    const regexp = await compile({
      foo: /[a-z]+/,
      bar: [
        /[0-9]+/,
        {
          bar: /[0-5]+/,
        },
      ],
    })
    const result = syntax(regexp, 'hello 123 789 world')
    expect(result).toMatchSnapshot()
  })
})
