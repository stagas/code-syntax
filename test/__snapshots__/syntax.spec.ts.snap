// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntax mapped regexp 1`] = `"<span class=\\"foo\\">hello</span> <span class=\\"bar-outer\\"><span class=\\"bar\\">123</span></span> <span class=\\"bar-outer\\">789</span> <span class=\\"foo\\">world</span>"`;

exports[`syntax multiple named group regexp 1`] = `"<span class=\\"foo\\">hello</span> <span class=\\"bar\\">123</span> <span class=\\"foo\\">world</span>"`;

exports[`syntax simple named group regexp 1`] = `"<span class=\\"foo\\">hello</span> 123 <span class=\\"foo\\">world</span>"`;
